### ISSUE TITLE
same augmentation for image and mask BUT different interpolation order

### ISSUE THREAD

**Issue by alexgais** [aleju/imgaug#135]
Hi Guys,
Great repo.

Can you share an example how I use same augmentation (and parameters values) for an image and a mask but implement different interpolation order - 1 for image, 0 for binary mask.

Thanks.

**Comment by aleju** [comment]
This should roughly be how different augmentation sequences can be used while also synchronizing augmenters/parameters in them (not tested):
```python
seq_img = iaa.Sequential([
    iaa.Add((-10, 10), name="MyAdd"),
    iaa.Affine(rotate=(-45, 45), order=1, mode=["constant", "edges"], cval=(0, 255), name="MyAffine")
])
seq_masks = iaa.Sequential([
    iaa.Affine(rotate=(-45, 45), order=0, mode="constant", cval=0, name="MyAffine")
])

for batch in batches:
    seq_masks_i = seq_masks.copy_random_state(seq_img, matching="name")
    seq_masks_i = seq_masks_i.to_deterministic()
    seq_img_i = seq_img.to_deterministic()

    imgs_aug = seq_img_i.augment_images(batch.images)
    masks_aug = seq_masks_i.augment_images(batch.masks.astype(np.uint8) * 255) >= 128
```

**Comment by ninadst** [comment]
I end up with an exception when I try to use this code

Exception: You called copy_random_state_() with a source that uses global random states. Call localize_random_state_() on the source first or initialize your augmenters with local random states, e.g. via Dropout(..., random_state=1234).

The entire code follows: 

```python
from imgaug import augmenters as iaa
from keras.preprocessing.image import array_to_img, img_to_array, load_img
import numpy as np 

seq_img = iaa.Sequential([
    iaa.Add((-10, 10), name="MyAdd"),
    iaa.Affine(rotate=(-45, 45), order=1, mode=["constant", "edges"], cval=(0, 255), name="MyAffine")
])

seq_masks = iaa.Sequential([
    iaa.Affine(rotate=(-45, 45), order=0, mode="constant", cval=0, name="MyAffine")
])


img = load_img('temp/img/1_1.bmp')  # this is a PIL image
x = img_to_array(img)  # this is a Numpy array with shape (3, 150, 150)
images = x.reshape((1,) + x.shape)  # this is a Numpy array with shape (1, 3, 150, 150)

img = load_img('temp/labels/1_1.bmp')  # this is a PIL image
y = img_to_array(img)  # this is a Numpy array with shape (3, 150, 150)
masks = y.reshape((1,) + y.shape)  # this is a Numpy array with shape (1, 3, 150, 150)

seq_masks_i = seq_masks.copy_random_state(seq_img, matching="name")
seq_masks_i = seq_masks_i.to_deterministic()
seq_img_i = seq_img.to_deterministic()

imgs_aug = seq_img_i.augment_images(images)
masks_aug = seq_masks_i.augment_images(masks.astype(np.uint8) * 255) >= 128

```

**Comment by ninadst** [comment]
I modified the code for my purpose. However, the image and masks not aligned properly.

```python
from imgaug import augmenters as iaa
from keras.preprocessing.image import array_to_img, img_to_array, load_img
from scipy import misc
import imageio 
import numpy as np

seq_img = iaa.Sequential([
    iaa.Affine(rotate=(-45, 45), order=1, mode="constant", cval=0, name="MyAffine")
])


seq_img = seq_img.localize_random_state()

seq_masks = iaa.Sequential([
    iaa.Affine(rotate=(-45, 45), order=0, mode="constant", cval=0, name="MyAffine")
])


img = load_img('temp/img/1_1.bmp')  # this is a PIL image
x = img_to_array(img)  # this is a Numpy array with shape (3, 150, 150)
images = x.reshape((1,) + x.shape)  # this is a Numpy array with shape (1, 3, 150, 150)

img = load_img('temp/labels/1_1.bmp')  # this is a PIL image
y = img_to_array(img)  # this is a Numpy array with shape (3, 150, 150)
masks = y.reshape((1,) + y.shape)  # this is a Numpy array with shape (1, 3, 150, 150)


i = 1
seq_masks_i = seq_masks.copy_random_state(seq_img, matching="name")
seq_masks_i = seq_masks_i.to_deterministic()
seq_img_i = seq_img.to_deterministic()
  
imgs_aug = seq_img_i.augment_images(images)
masks_aug = seq_masks_i.augment_images(masks)
    
imageio.imwrite("img_%06d.bmp"% (i,), np.squeeze(imgs_aug))
imageio.imwrite("msk_%06d.bmp"% (i,), np.squeeze(masks_aug))
 
```

**Comment by aleju** [comment]
Can't check in detail without your images test, but the library expects the channel-axis to be at the end of the array, i.e. (N, H, W, C) and not (N, C, H, W), which you have according to your comments.
Also, place `copy_random_state()` after the two `to_deterministic()` calls as they randomize the internal random state.

**Comment by ninadst** [comment]
@aleju Thanks for your reply. I was able to apply the same augmentation to both the image and corresponding mask after fixing the code based on your suggestion. I am leaving self-contained code below for reference.The motivation behind this code is to apply nearest neighbor interpolation to the mask so that no new values are introduced to the mask.
```python
from imgaug import augmenters as iaa
from keras.preprocessing.image import array_to_img, img_to_array, load_img
from scipy import misc
import imageio 
import numpy as np
from PIL import Image, ImageDraw

seq_img = iaa.Sequential([
    iaa.Affine(rotate=(-45, 45), order=1, mode="constant", cval=0, name="MyAffine")
])

seq_masks = iaa.Sequential([
    iaa.Affine(rotate=(-45, 45), order=0, mode="constant", cval=0, name="MyAffine")
])


img = Image.new('L', [400,400])
d = ImageDraw.Draw(img)
d.rectangle([(150,150),(250,250)], fill=255)

x = img_to_array(img) 
images = x.reshape((1,) + x.shape)  

masks =images

seq_img = seq_img.localize_random_state()

seq_img_i = seq_img.to_deterministic()
seq_masks_i = seq_masks.to_deterministic()

seq_masks_i = seq_masks_i.copy_random_state(seq_img_i, matching="name")
  
imgs_aug = seq_img_i.augment_images(images)
masks_aug = seq_masks_i.augment_images(masks)
    
imageio.imwrite("img.bmp", np.squeeze(imgs_aug))
imageio.imwrite("msk.bmp", np.squeeze(masks_aug))
